### PR TITLE
References support

### PR DESCRIPTION
**Pull request by dcolligan**
Issue #374

### CODE REVIEW

**Comment by dcolligan** [comment]
submitting a PR since I'm sure @jeromekelleher will have thoughts on the code thus far

**Comment by jeromekelleher** [comment]
Frankly, I'm a bit in the dark @dcolligan. I don't think it's possible to turn an arbitrary collection of FASTA files into a GA4GH ReferenceSet, so we're going to have to add some metadata. I don't have any experience with FASTA, so I don't have any idea of how this should be done.

@diekhans have you any thoughts here? What's the right way to do this? A single FASTA file corresponding to a ReferenceSet, which contains a bunch of sequences, or a directory full of files, each containing 1 or more sequences? How do we specify the required metadata? Do we attempt to decipher the metadata line in the FASTA file, or do we just require the existence of a 'metadata.json' file (for example) holding the information that we need?

Anyone else want to weigh in here?

**Comment by dcolligan** [comment]
Trying to parse the first line of a FASTA file for metadata is madness.  There's absolutely no standardization there, and if we invent our own ad-hoc format it should be at least a bit more structured.

**Comment by diekhans** [comment]
> Trying to parse the first line of a FASTA file for metadata is madness.

Welcome to bioinformatics ;-)

This is one of the huge problems GA4GH is trying to fix.

I think the right approach is to have libraries handle the common
work and then have commands tailored to specific sources of data.

Even though GFF3 is better standardized than FASTA, it's going
to be far easier to write a gencode_gff3 import command than
works with a specific data set than tring to have some that is
parameterized to handle special cases semantics.  I would rather
have lots of specific commands than complex `soft coding' of a
general one.

We have libraries and people provide commands to get their 
data to the right format.

**Comment by jeromekelleher** [comment]
Some specific questions for you above @diekhans...

**Comment by pgrosu** [comment]
It's not as bad as people think.  Standardization is your friend and yes @diekhans is right :)  We fixed this problem in industry by mapping to controlled vocabulary and sometimes restructuring it.  Anyone submitting references would pre-process them before submitting them to the server to integrate with other data.

**Comment by dcolligan** [comment]
@jeromekelleher pushed an updated change.  Want your opinion on how we should abstract the stack of the listReferenceBases code path, or if it's worth it at all for just one server method.

**Comment by jeromekelleher** [comment]
I had a quick look at the `listReferenceBases` code path @dcolligan, and not crazy about the current setup. My main concern is the flask dependency within `backend.py`: I think backend.py should be able to operate completely independently of Flask.  I think we should try to follow the pattern of the rest of the objects, and create a dictionary out of the Flask params within frontend.py, and then pass this on to the appropriate handler in the backend where it can be checked to see if it is an instance of  `ListReferenceBasesRequest`. This is slightly different to the search requests (where we pass the data as a string), but I think this is OK.

**Comment by dcolligan** [comment]
@jeromekelleher alright, pushed an update; let me know what you think

File: ga4gh/frontend.py [hunk 1]
```diff
@@ -207,6 +214,24 @@ def handleHttpPost(request, endpoint):
     return getFlaskResponse(responseStr)
 
 
+def handleList(id_, endpoint, request):
+    """
+    Handles the specified HTTP GET request, mapping to a list request
+    """
+    requestArgs = request.args
```
Comment (jeromekelleher) [inline ga4gh/frontend.py]: Might as well just include `request.args` in the list of parameters here.

File: ga4gh/backend.py [hunk 2]
```diff
@@ -209,6 +214,83 @@ def getReadGroupSets(self):
...
+        return self.runGetRequest(self._referenceSetIdMap, id_)
+
+    def listReferenceBases(self, id_, requestArgs):
+        # parse arguments
+        try:
+            reference = self._referenceIdMap[id_]
+        except KeyError:
+            raise exceptions.ObjectWithIdNotFoundException(id_)
+        start = 0
+        end = datamodel.PysamDatamodelMixin.fastaMax
```
Comment (jeromekelleher) [inline ga4gh/backend.py]: Is there anyway that we can use the validation machinery here to avoid some of this messiness? Can't we create a `ListReferenceBasesRequest` object out of the requestArgs dictionary? We then have full validation, we reuse the general machinery for parsing and life is good.

**Comment by jeromekelleher** [comment]
Generally looks good @dcolligan, modulo the minor comments above.  I'll need to spend some time to play with it properly and go through the code more thoroughly though.

File: ga4gh/backend.py (hunk 2 above)
Comment (dcolligan) [inline ga4gh/backend.py]: `protocol.fromJsonDict` just assigns a value if there is an attribute name match; it doesn't check the type.  (Not sure if we want to change this.)

That is, we need to cast `requestArgs.start` to an int from a string ourselves.

Comment (jeromekelleher) [inline ga4gh/backend.py]: Ah right; we can't call `validate` on the dict as it's just a string:string map, and there's no type conversion done. OK, if there are no other GET requests with multiple typed arguments (are there, or will there be?) then I guess what you have is as good as we need. If there are, then I think it's worthwhile thinking about how to generalise this code, and getting automated parsing and type checking of these values from the schema. We can open an issue on this if you would prefer not to slow this PR down any more.

Comment (dcolligan) [inline ga4gh/backend.py]: Yeah, I'm generally an advocate of "incremental progress with less-than-perfect code, and refactor as it becomes a problem" over "implement clean, very general abstractions before they are necessarily widely needed."  (As has been said in other contexts before, the later is so... java.)  Let's go forward with this PR and file and issue to generalize list request abstractions.

There is still this puzzling CI failure to fix, though.

**Comment by jeromekelleher** [comment]
I've had a quick play with this @dcolligan, and I think there are still some issues:

1. When I run `referencesets-search` on the client, the `referenceIds` attribute is empty in the returned

File: tests/datadriven/test_references.py [hunk 3]
```diff
@@ -31,23 +32,89 @@ class ReferenceSetTest(datadriven.DataDrivenTest):
...
+        Container class for information about a reference
+        """
+        def __init__(self, referenceSetId, fastaFileName):
+            filename = os.path.split(fastaFileName)[1]
             referenceId = "{}:{}".format(
                 referenceSetId, filename.split(".")[0])
-            self._idFastaFileMap[referenceId] = pysam.FastaFile(path)
+            self.id = referenceId
+            self.fastaFile = pysam.FastaFile(fastaFileName)
+            self.bases = self.fastaFile.fetch(self.fastaFile.references[0])
```
Comment (jeromekelleher) [inline tests/datadriven/test_references.py]: Is this reading all the bases into memory? If so, we can't do this. GRCH37 is 3GB...

Comment (dcolligan) [inline tests/datadriven/test_references.py]: Well, let's not include GRCH37 in our test data files, then!

Comment (jeromekelleher) [inline tests/datadriven/test_references.py]: Oops, sorry! My bad, didn't realise this was the test path. This brings up another issue: shouldn't we be using some other library to access the FASTA data (i.e. not pysam) as a sanity check?

Comment (dcolligan) [inline tests/datadriven/test_references.py]: Maybe, but it's no worse than what we do in other test modules :)

**Comment by dcolligan** [comment]
Alright, I finally fixed the CI (the issue was that the CI was passing locally but failing on travis because the `.gz` file was not checked in and therefore not present when the test was run on travis, but that wasn't obvious to me since `*.gz` is a pattern in `.gitignore`).

File: tests/datadriven/test_references.py (hunk 3 above)
Comment (jeromekelleher) [inline tests/datadriven/test_references.py]: Only for reads, and there because we have to essentially. For variants we use PyVCF.

**Comment by jeromekelleher** [comment]
What's the status with this @dcolligan? Have you tried it with large-scale data? If so, I'll have another play with it and hopefully we can get it merged in the next couple of days.

**Comment by dcolligan** [comment]
@jeromekelleher `references-get` works for me.  Maybe you're forgetting to pass in an id (it's now a required field for that command, not an optional field -- I changed it because this seems to make more sense to error out early on the client)

**Comment by dcolligan** [comment]
@jeromekelleher pushed a new update that fixes `referencesets-search` not having the `referenceIds` attribute set

**Comment by dcolligan** [comment]
@jeromekelleher that ftp link you posted above doesn't seem to be responsive... do you have any other examples of "real world data"?  also, what other FASTA files do you think we should add to the datadriven tests?  Unless there are some corner cases we're covering, I don't really see the value...

**Comment by jeromekelleher** [comment]
OK, the NCBI server is on your side of the water and should be accessible: ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/phase2_reference_assembly_sequence/. There's some real-world references here, which we're going to need to support soon, so it'll be good to get some experience of working with them.

**Comment by dcolligan** [comment]
Ok, I loaded the NCBI reference (~3G uncompressed) into the reference server and it seems to work.  The only thing I noticed is that startup of the server is pretty slow, which is due to generating the MD5 checksums.  Precomputing these would definitely speed things up, but I'm not sure if we want to go down that road.  Thoughts?

I needed to split the given NCBI FASTA file up into several in order to get it to map to our model.  See `scripts/split_fasta.py` for a tool I added to help do that.

**Comment by jeromekelleher** [comment]
Excellent, thanks @dcolligan. For now, I would disable md5 sums --- we'll need to store these, and a bunch of other metadata about the references, so we might as well do that all in one go in a follow up. Recomputing md5s on 3G files each time the server starts up would be very frustrating for users, not to mention bad for scale-out performance, where we want to be able to quickly spin up new instances.

Do we really need to split the FASTA up? Can't we reuse the convention from VCFs, where we have one or more files, and we automatically map to the file in question based on the query?

**Comment by dcolligan** [comment]
@jeromekelleher just pushed a change that disabled Md5 computation.  Server startup is much faster.

Re: FASTA file splitup: The `ListReferenceBasesRequest` request is to `GET /references/{id}/bases` and has three attributes: `end`, `start`, and `pageToken`.  There is no attribute for indexing into a specific sequence within a reference.

**Comment by jeromekelleher** [comment]
I think we can work around that issue easily enough @dcolligan, but we should merge this much now. It's arguable whether we should support single fasta references or not: if someone really wants it, let them open an issue on it.

This looks good and works well; thanks @dcolligan. It's starting to look like a real server now!

+1.
